### PR TITLE
A new filter "ifindex" for LINUX_SLL2 and live Linux captures

### PR DESCRIPTION
**Pull request by fenner**
Create a filter for the ifIndex field in the LINUX_SLL2 pcap format, and convert it to the right SKF_AD_ value for live captures.

### CODE REVIEW

**Comment by fxlb** [comment]
As a user, I’d rather be able to filter by interface name like in some BSD's but in and out.

Something like `ifname eth0`.

Way to have 'eth0' -> 'ifindex N' to build a filter?

See https://www.tcpdump.org/manpages/pcap-filter.7.html:

"ifname interface 
True if the packet was logged as coming from the specified interface (applies only to packets logged by OpenBSD's or FreeBSD's pf(4))."

**Comment by fenner** [comment]
> As a user, I’d rather be able to filter by interface name like in some BSD's but in and out.

You can't do this on a savefile since the interface name is not stored in the savefile.  ifindex works in both live captures and savefile mode, even if it's less friendly.

Reworking the grammar so that "ifname" is not PF-specific is on my to-do list, but that'd be a different pull request.

**Comment by mcr** [comment]
> > As a user, I’d rather be able to filter by interface name like in some BSD's but in and out.
> 
> You can't do this on a savefile since the interface name is not stored in the savefile. ifindex works in both live captures and savefile mode, even if it's less friendly.
> 
> Reworking the grammar so that "ifname" is not PF-specific is on my to-do list, but that'd be a different pull request.

And even if you were on the same machine, the ifindex can change after interfaces come/go, or after a reboot.    This is a case where pcapng would be better.

**Comment by fxlb** [comment]
I suggest to rebase this PR on top of master.

**Comment by fenner** [comment]
Looks like 8c263fd introduced problems with the FreeBSD build.

**Comment by guyharris** [comment]
> Looks like [8c263fd](https://github.com/the-tcpdump-group/libpcap/commit/8c263fda3f8f274359ea47ecf667094443c704c9) introduced problems with the FreeBSD build.

Should be fixed by 00f29bec47f53e747c188f3b23aa7b0c4d45db6e.

File: gencode.c [hunk 1]
```diff
@@ -8175,6 +8175,53 @@ gen_multicast(compiler_state_t *cstate, int proto)
...
+	if (setjmp(cstate->top_ctx))
+		return (NULL);
+
+	/*
+	 * Only some data link types support ifindex qualifiers.
+	 */
+	switch (cstate->linktype) {
+	case DLT_LINUX_SLL2:
+		/* match packets on this interface */
+		b0 = gen_cmp(cstate, OR_LINKHDR, 4, BPF_W, ifindex);
```
Comment (guyharris) [inline gencode.c:8196]: DLT_LINUX_SLL and DLT_LINUX_SLL2 headers are generated by libpcap; they're not visible to the in-kernel filter engine.

This probably needs to be done with something such as

    if (doing a live capture) {
        #ifdef linux
            test SKF_AD_OFF + SKF_AD_IFINDEX
        #else
            report "only supported on Linux"
        #endif
    } else {
        if (linktype is DLT_LINUX_SLL2)
            test the field at an offset of 4
        else
            report "not supported on savefiles except for DLT_LINUX_SLL2
    }

Comment (fenner) [reply gencode.c:8196]: > DLT_LINUX_SLL and DLT_LINUX_SLL2 headers are generated by libpcap; they're not visible to the in-kernel filter engine.

Right, that's why `pcap-linux.c:fix_offset()` changes the filter on the fly when it's `DLT_LINUX_SLL*`.

> This probably needs to be done with something such as

I think my code is equivalent to what you suggest, but has the advantage of having strong parallels with the existing `gen_inbound()`.  Note that if we're doing a live capture, the code in `pcap-linux.c:fix_offset()` will turn the offset 4 comparison into an offset `SKF_AD_OFF + SKF_AD_IFINDEX` comparison.

Comment (guyharris) [reply gencode.c:8196]: OK.

File: gencode.c [hunk 2]
```diff
@@ -8175,6 +8175,53 @@ gen_multicast(compiler_state_t *cstate, int proto)
...
+		break;
+        default:
+#if defined(linux) && defined(PF_PACKET) && defined(SO_ATTACH_FILTER)
```
Comment (guyharris) [inline gencode.c:8199]: At this point, if we're building for Linux and neither `PF_PACKET` nor `SO_ATTACH_FILTER` are defined, libpcap won't build - we require a newer kernel - so it's sufficient to test whether `linux` is defined.

Comment (fenner) [reply gencode.c]: Sure, I first posted this way before bfbe856 . Thanks for pointing this out.

**Pull request merged**